### PR TITLE
Corrects the mapping of types to features in ITypeFeatureProvider 

### LINKED ISSUE #15782: DI dependencies of features in the same assembly are mapped to the wrong feature
### Describe the bug
In case a single assembly contains several features, the dependencies added to the `ITypeFeatureProvider` may not be correct. All dependencies are added to one arbitrary feature, no matter which feature actually registers it in DI. I guess anything is added to the first feature being enabled.

In turn this causes things like the `RoleUpdater` to not correctly apply default permissions, because enabling a second feature in a single module is not correctly detected.

This can only be prevented, when all necessary dependencies are marked with the `FeatureAttribute`.

### PR DESCRIPTION
**Pull request by gvkries**
Just an idea how to fix #15782. Removing `FeatureEntry` may be overkill, but it is not really required.

### CODE REVIEW

**Comment by Piedone** [comment]
This may substitute https://github.com/OrchardCMS/OrchardCore/pull/15787.

**Comment by MikeAlhayek** [comment]
@gvkries what actually fixes the issue here?

**Comment by gvkries** [comment]
> @gvkries what actually fixes the issue here?

It's in the `TypeFeatureProvider.TryAdd` method.

**Comment by MikeAlhayek** [comment]
@gvkries I think there is two things that should be considered in the solutions.

1. A Type can belong to one or more feature.
2. I think the solution should be part of the `ShellContainerFactory`.

The only way to really know if a type was added by a startup file, is by calling that start up and then get all the types that were added in the process. I am not sure if this is the best process 

I think something like this may actually solve the problem of needed `[Feature]` attribute all over the place. With this approach we can remove the calling `typeFeatureProvider.TryAdd` from the extension manager.

```
foreach (var startup in startups)
{
    var feature = blueprint.Dependencies.FirstOrDefault(x => x.Key == startup.GetType()).Value?.FeatureInfo;

    startup.ConfigureServices(featureServiceCollection);

    foreach (var type in featureServiceCollection.Select(x => x.GetImplementationType()))
    {
        if (type is null)
        {
            continue;
        }

        typeFeatureProvider.TryAdd(type, feature ?? _applicationFeature);
    }
}
```
With this approach, only a startup class will need a `[Feature]` attribute. Any class that is registered in that startup will automatically be tied to it. But then we also have to change the implementation of `ITypeFeatureProvider` so it allows mapping many features to a type.

I am not sure I have time to test out the above approach, but if you are able it, it may worth trying.

**Comment by gvkries** [comment]
@MikeAlhayek I take a look if I find some more time. But do you think we should work on this at all, or is it too much hassle?

**Comment by MikeAlhayek** [comment]
@gvkries I think we should. But we have to replace existing approach with a better approach. The current solution is to use `Feature[]` attribute. And we agree that it's not the best approach. So we need to come up with a better solution. If the solution I suggested works, then I think this will be win win and the developer will not need to use the Feature attribute outside the startup and the controller.

**Comment by gvkries** [comment]
@MikeAlhayek Your solution is almost the same as mine, I only kept the type registration in the `ExtensionManager` as-is and removed the `FeatureEntry`. `FeatureEntry` purely exists to keep a list of public types per feature (which is then also not correctly mapped to features without the `FeatureAttribute`). It has only two uses, one in the `CompositionStrategy` and one in the `RoleUpdater`. The role updater must be changed anyway to use the `ITypeFeatureProvider` instead, which leaves us with a single use case for the `FeatureEntry` during startup only. Additionally it keeps the type lists alive for the app lifetime, which is rarely or never used again (only during tenant restarts I think).
Maybe the best solution is to rework the `CompositionStrategy` as well. This way we may can get rid of the type mappings being added by the extension manager at all.

**Comment by gvkries** [comment]
I tried this quickly, but `ShellContainerFactory` comes into the game too late, the `ITypeFeatureProvider` is already used before. So removing it being filled by the `ExtensionManager` brings some challenges.

I need to go now, I'll come back to this later.

**Comment by MikeAlhayek** [comment]
Thank for looking into it.

maybe we need to investigate why do we need to populate the `_typeFeatureProvider` using the extension manager? Quick look at the code tells me what should not have to do it that early. This is something that we need to do during the startup process. But I could be wrong since I have not really had time to peal this onion.

**Comment by gvkries** [comment]
> I tried this quickly, but `ShellContainerFactory` comes into the game too late, the `ITypeFeatureProvider` is already used before. So removing it being filled by the `ExtensionManager` brings some challenges.

My reasoning was wrong, I only missed some types in the feature provider.

**Comment by MikeAlhayek** [comment]
What we build a fake Service Collection in the Extension Manager so we can get the correct mapping and therefore we won't need it in the ShellContainerFactory.

**Comment by gvkries** [comment]
As @MikeAlhayek said, I completely moved populating the `ITypeFeatureProvider` into the `ShellContainerFactory`. The shell container factory can then associate all types to their features according to their startups and also adds all types that are not held by DI.

Additionally this simplifies the extension manager and moves all usages that needs the feature types to rely on the type feature provider as well. No more duplicated responsibilities between extension manager and type feature provider.

Type sharing between multiple features is not considered yet.

**Comment by gvkries** [comment]
Yes, I think we should skip that addition for now.

**Comment by gvkries** [comment]
> > Yes, I think we should skip that addition for now.
> 
> I am not sure you can. The extension manager uses every feature available (not the features that are associated to the current tenant). This means that if you have a type **T** that could be registered to feature **A** and feature **B** which both features are independent of each other, then the extension manager may assign type **T** to feature **A** and never to feature **B**. Now if you only enable feature **B** and not **A**, then the `ITypeFeatureProvider` will contain one less type which is what you are trying to solve here.

Yeah, that is correct. But the current solution is already an improvement.

**Comment by MikeAlhayek** [comment]
Not really. Using the Feature attribute currently gives you the correct mapping no matter what. If we make this change, then Feature attribute will not be honored as we would honer registration. In order for this to be an acceptable solution, I think it has to provide a better replacement to the existing solution.

**Comment by gvkries** [comment]
No, the current solution does not remove or change the feature attribute at all. Only types not marked with it are now correctly assigned to its feature if it is in the DI container. I was not really thinking of fully replacing the `FeatureAttribute`, only improving the assignment to avoid errors that are hard to find. But it's still work-in-progress, let me see what I can do with that problem...

**Comment by gvkries** [comment]
I've changed the `ITypeFeatureProvider` to allow types that are used by multiple features. I think this is mostly correct, but there are some challenges still to be done. E.g. all callees of `ITypeFeatureProvider.GetFeaturesForDependency()` will have to consider the fact that a returned feature may not be enabled at all.

**Review by Piedone — COMMENTED** [review]
Sorry that I didnt join earlier. I'm still trying to understand this, but great discussion and work! It was quite a ride to go through all of this :).

Would you be able to join the Thursday triage meeting (https://docs.orchardcore.net/en/latest/resources/meeting/) and present this? We need a group consultation.

I think this approach is good, but I need to put more time into it.

**Comment by gvkries** [comment]
> Would you be able to join the Thursday triage meeting (https://docs.orchardcore.net/en/latest/resources/meeting/) and present this? We need a group consultation.

I think I am able to join next week.

File: src/OrchardCore.Modules/OrchardCore.Roles/Controllers/AdminController.cs
[the remark this answers is not in the document]
Comment (gvkries) [reply src/OrchardCore.Modules/OrchardCore.Roles/Controllers/AdminController.cs:1]: Not necessarily, applying the feature attribute should still be possible and the behavior should not be changed by this PR.

File: src/OrchardCore/OrchardCore.Abstractions/Extensions/Features/ITypeFeatureProvider.cs [hunk 1]
```diff
@@ -9,7 +10,8 @@ namespace OrchardCore.Environment.Extensions
     /// </summary>
     public interface ITypeFeatureProvider
     {
-        IFeatureInfo GetFeatureForDependency(Type dependency);
+        IEnumerable<IFeatureInfo> GetFeaturesForDependency(Type dependency);
```
[the remark this answers is not in the document]
Comment (gvkries) [reply src/OrchardCore/OrchardCore.Abstractions/Extensions/Features/ITypeFeatureProvider.cs:13]: Yes, that is a valid idea. But maybe we should review the responsibilities of `ITypeFeatureProvider` and `IExtensionManager` completely, there is some overlapping.

**Comment by Piedone** [comment]
That would be great, thank you!

File: src/OrchardCore.Modules/OrchardCore.Roles/Controllers/AdminController.cs
[the remark this answers is not in the document]
Comment (Piedone) [reply src/OrchardCore.Modules/OrchardCore.Roles/Controllers/AdminController.cs:1]: While it will work, it'll have an unnecessary attribute.

**Comment by Piedone** [comment]
I couldn't take part today, but did you in the end?

**Comment by gvkries** [comment]
No, sorry. I was talking about this week, so next Thursday 👍

**Comment by Piedone** [comment]
OK, I'll  try to take part too.

File: src/OrchardCore/OrchardCore.Abstractions/Extensions/Features/ITypeFeatureProvider.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (gvkries) [reply src/OrchardCore/OrchardCore.Abstractions/Extensions/Features/ITypeFeatureProvider.cs:32]: I added `GetExtensionForDependency`, because that makes sense as a replacement for some calls to `GetFeatureForDependency`. I also added the original `GetFeatureForDependency` back, for cases where only a single feature makes sense. This makes using the `ITypeFeatureProvider` easier.

[the remark this answers is not in the document]
Comment (Piedone) [reply src/OrchardCore/OrchardCore.Abstractions/Extensions/Features/ITypeFeatureProvider.cs:32]: Great!

**Comment by Piedone** [comment]
It's now under http://orchardcore.net/meeting BTW :).

**Review by sebastienros — APPROVED** [review]

File: src/OrchardCore.Modules/OrchardCore.Roles/Controllers/AdminController.cs
[the remark this answers is not in the document]
Comment (Piedone) [reply src/OrchardCore.Modules/OrchardCore.Roles/Controllers/AdminController.cs:1]: Remove the attribute now there, please.

File: src/OrchardCore/OrchardCore.Abstractions/Extensions/Features/ITypeFeatureProvider.cs [hunk 2]
```diff
@@ -10,8 +10,39 @@ namespace OrchardCore.Environment.Extensions
     /// </summary>
     public interface ITypeFeatureProvider
     {
+        /// <summary>
+        /// Gets the extension for the specified dependent type.
+        /// </summary>
+        /// <param name="dependency"></param>
+        /// <returns></returns>
```
Comment (Piedone) [inline src/OrchardCore/OrchardCore.Abstractions/Extensions/Features/ITypeFeatureProvider.cs:17]: Remove empty `<returns></returns>`s and empty parameter docs.

File: src/OrchardCore/OrchardCore.Abstractions/Extensions/Features/ITypeFeatureProvider.cs [hunk 3]
```diff
@@ -10,8 +10,39 @@ namespace OrchardCore.Environment.Extensions
     /// </summary>
     public interface ITypeFeatureProvider
     {
+        /// <summary>
+        /// Gets the extension for the specified dependent type.
```
Comment (Piedone) [inline src/OrchardCore/OrchardCore.Abstractions/Extensions/Features/ITypeFeatureProvider.cs:14]: Add a `<remark>` about why the first one is interesting, why do we have this separate methods (because otherwise it's curious: "what's the significance of the first feature?").

File: src/OrchardCore/OrchardCore.Mvc.Core/ModularApplicationModelProvider.cs [hunk 4]
```diff
@@ -44,25 +44,22 @@ public void OnProvidersExecuted(ApplicationModelProviderContext context)
             {
                 var controllerType = controller.ControllerType.AsType();
 
-                var blueprint = _typeFeatureProvider.GetFeaturesForDependency(controllerType).FirstOrDefault();
+                var blueprint = _typeFeatureProvider.GetExtensionForDependency(controllerType);
 
-                if (blueprint != null)
```
Comment (Piedone) [inline src/OrchardCore/OrchardCore.Mvc.Core/ModularApplicationModelProvider.cs:47]: I don't know if this null check is necessary, but only remove it if you're sure it isn't.

File: src/OrchardCore/OrchardCore.Mvc.Core/ModularApplicationModelProvider.cs [hunk 5]
```diff
@@ -44,25 +44,22 @@ public void OnProvidersExecuted(ApplicationModelProviderContext context)
...
-                { 
-                    if (blueprint.Extension.Id == _hostingEnvironment.ApplicationName && !_shellSettings.IsRunning())
+                if (blueprint.Id == _hostingEnvironment.ApplicationName && !_shellSettings.IsRunning())
+                {
+                    // Don't serve any action of the application'module which is enabled during a setup.
```
Comment (Piedone) [inline src/OrchardCore/OrchardCore.Mvc.Core/ModularApplicationModelProvider.cs:51]:
```suggestion
                    // Don't serve any action of the application's module which is enabled during a setup.
```

File: src/OrchardCore/OrchardCore/Extensions/Features/TypeFeatureProvider.cs [hunk 6]
```diff
@@ -10,6 +10,28 @@ public class TypeFeatureProvider : ITypeFeatureProvider
...
+        public IFeatureInfo GetFeatureForDependency(Type dependency)
+        {
+            if (_features.TryGetValue(dependency, out var features))
+            {
+                // Gets the first feature that has the same ID as the extension, if any. Otherwise returns the
+                // first feature.
+                return features.FirstOrDefault(feature => feature.Extension.Id == feature.Id) ?? features.First();
+            }
+
+            throw new InvalidOperationException($"Could not resolve main feature for type {dependency.Name}");
```
Comment (Piedone) [inline src/OrchardCore/OrchardCore/Extensions/Features/TypeFeatureProvider.cs:32]:
```suggestion
            throw new InvalidOperationException($"Could not resolve extension for type {dependency.Name}.");
        }

        public IFeatureInfo GetFeatureForDependency(Type dependency)
        {
            if (_features.TryGetValue(dependency, out var features))
            {
                // Gets the first feature that has the same ID as the extension, if any. Otherwise returns the
                // first feature.
                return features.FirstOrDefault(feature => feature.Extension.Id == feature.Id) ?? features.First();
            }

            throw new InvalidOperationException($"Could not resolve main feature for type {dependency.Name}.");
```

**Review by Piedone — CHANGES_REQUESTED** [review]

File: src/OrchardCore/OrchardCore.Mvc.Core/ModularApplicationModelProvider.cs (hunk 4 above)
Comment (gvkries) [reply src/OrchardCore/OrchardCore.Mvc.Core/ModularApplicationModelProvider.cs:47]: The `TypeFeatureProvider` will throw if no extension is available.

**Comment by MikeAlhayek** [comment]
@gvkries can you please confirm that we solved the issue where one type can belong to multiple features?

Also, at this point you should also remove the `[Feature]` attribute from classes like Migration, shape table provider, permissions (anything except for Startup files and controllers).

File: src/OrchardCore/OrchardCore.Mvc.Core/ModularApplicationModelProvider.cs (hunk 4 above)
Comment (Piedone) [reply src/OrchardCore/OrchardCore.Mvc.Core/ModularApplicationModelProvider.cs:47]: Indeed.

File: src/OrchardCore/OrchardCore.Mvc.Core/ModularApplicationModelProvider.cs [hunk 7]
```diff
@@ -44,25 +44,22 @@ public void OnProvidersExecuted(ApplicationModelProviderContext context)
             {
                 var controllerType = controller.ControllerType.AsType();
 
-                var blueprint = _typeFeatureProvider.GetFeaturesForDependency(controllerType).FirstOrDefault();
+                var blueprint = _typeFeatureProvider.GetExtensionForDependency(controllerType);
```
Comment (Piedone) [inline src/OrchardCore/OrchardCore.Mvc.Core/ModularApplicationModelProvider.cs:47]:
```suggestion
                var extension = _typeFeatureProvider.GetExtensionForDependency(controllerType);
```

**Comment by Piedone** [comment]
Yes, it's a different time: https://docs.orchardcore.net/en/latest/resources/meeting/

**Comment by gvkries** [comment]
> @gvkries can you please confirm that we solved the issue where one type can belong to multiple features?
> 
> Also, at this point you should also remove the `[Feature]` attribute from classes like Migration, shape table provider, permissions (anything except for Startup files and controllers).

Yes, all types can be assigned to as one or more features. The `ITypeFeatureProvider` returns an enumerable of features now. But we have several occasions where only one feature is expected, that part is a little bit tricky. The provider will return the first feature which has the same ID as the extension in that case. This mimics the previous behavior, because the extension manager has always assigned the type to the feature with the extension ID there too.

I will remove the feature attribute and check any problem arises from that.

File: src/OrchardCore/OrchardCore/Extensions/Features/TypeFeatureProvider.cs [hunk 8]
```diff
@@ -1,26 +1,55 @@
...
+            if (_features.TryGetValue(dependency, out var features))
             {
-                return feature;
+                // Gets the first feature that has the same ID as the extension, if any. Otherwise returns the
+                // first feature.
+                return features.FirstOrDefault(feature => feature.Extension.Id == feature.Id) ?? features.First();
             }
 
-            throw new InvalidOperationException($"Could not resolve feature for type {dependency.Name}");
+            throw new InvalidOperationException($"Could not resolve main feature for type {dependency.Name}");
```
Comment (MikeAlhayek) [inline src/OrchardCore/OrchardCore/Extensions/Features/TypeFeatureProvider.cs:32]: shouldn't we just return null instead of throwing exception here?

File: src/OrchardCore/OrchardCore/Extensions/Features/TypeFeatureProvider.cs [hunk 9]
```diff
@@ -1,26 +1,55 @@
...
+        }
+
+        public IEnumerable<IFeatureInfo> GetFeaturesForDependency(Type dependency)
+        {
+            if (_features.TryGetValue(dependency, out var features))
+            {
+                return features;
+            }
+
+            throw new InvalidOperationException($"Could not resolve features for type {dependency.Name}");
```
Comment (MikeAlhayek) [inline src/OrchardCore/OrchardCore/Extensions/Features/TypeFeatureProvider.cs:42]: same here. probably null is better

File: src/OrchardCore/OrchardCore/Extensions/Features/TypeFeatureProvider.cs [hunk 10]
```diff
@@ -1,26 +1,55 @@
...
+        }
+
+        public IEnumerable<Type> GetTypesForFeature(IFeatureInfo feature)
+        {
+            return _features.Where(kv => kv.Value.Contains(feature)).Select(kv => kv.Key);
```
Comment (MikeAlhayek) [inline src/OrchardCore/OrchardCore/Extensions/Features/TypeFeatureProvider.cs:47]: materialize this using `.ToArray()` to prevent evaluating this every time

File: src/OrchardCore/OrchardCore/Shell/Builders/ShellContainerFactory.cs [hunk 11]
```diff
@@ -132,9 +132,62 @@ public async Task<IServiceProvider> CreateContainerAsync(ShellSettings settings,
...
+            PopulateTypeFeatureProvider(typeFeatureProvider, featureAwareServiceCollection);
+
+            return shellServiceProvider;
+        }
+
+        private void EnsureApplicationFeature()
+        {
+            if (_applicationFeature is null)
+            {
+                lock (this)
```
Comment (MikeAlhayek) [inline src/OrchardCore/OrchardCore/Shell/Builders/ShellContainerFactory.cs:145]: shouldn't the lock be for `_applicationFeature` instead of `this`? Then since `_applicationFeature` is locked, I don't think we need `??=` equals would be enough.

File: src/OrchardCore/OrchardCore/Shell/Builders/ShellContainerFactory.cs [hunk 12]
```diff
@@ -132,9 +132,62 @@ public async Task<IServiceProvider> CreateContainerAsync(ShellSettings settings,
...
+        }
+
+        private void PopulateTypeFeatureProvider(ITypeFeatureProvider typeFeatureProvider, FeatureAwareServiceCollection featureAwareServiceCollection)
+        {
+            // Get all types from all extension and add them to the type feature provider.
+            var extensions = _extensionManager.GetExtensions();
+
+            var allTypesByExtension = extensions
+                .SelectMany(extension =>
+                    _extensionManager.GetExportedExtensionTypes(extension).Where(IsComponentType)
```
Comment (MikeAlhayek) [inline src/OrchardCore/OrchardCore/Shell/Builders/ShellContainerFactory.cs:160]:
```suggestion
                    _extensionManager.GetExportedExtensionTypes(extension)
                    .Where(IsComponentType)
```

File: src/OrchardCore/OrchardCore/Shell/Builders/ShellContainerFactory.cs [hunk 13]
```diff
@@ -132,9 +132,62 @@ public async Task<IServiceProvider> CreateContainerAsync(ShellSettings settings,
...
+                    typeByExtension.Extension.Id))
+                .ToDictionary(
+                    group => group.Key,
+                    group => group.Select(typesByExtension => typesByExtension.Type));
+
+            foreach (var extension in extensions)
+            {
+                foreach (var feature in extension.Features)
+                {
+                    // Features can have no types
```
Comment (MikeAlhayek) [inline src/OrchardCore/OrchardCore/Shell/Builders/ShellContainerFactory.cs:179]:
```suggestion
                    // Features can have no types.
```

File: src/OrchardCore/OrchardCore/Shell/Builders/ShellContainerFactory.cs [hunk 14]
```diff
@@ -132,9 +132,62 @@ public async Task<IServiceProvider> CreateContainerAsync(ShellSettings settings,
...
+                    {
+                        foreach (var type in featureTypes)
+                        {
+                            typeFeatureProvider.TryAdd(type, feature);
+                        }
+                    }
+                }
+            }
 
+            // Register all DIed types in ITypeFeatureProvider
```
Comment (MikeAlhayek) [inline src/OrchardCore/OrchardCore/Shell/Builders/ShellContainerFactory.cs:190]:
```suggestion
            // Register all DIed types in ITypeFeatureProvider.
```

File: src/OrchardCore/OrchardCore/Extensions/Features/TypeFeatureProvider.cs (hunk 8 above)
Comment (gvkries) [reply src/OrchardCore/OrchardCore/Extensions/Features/TypeFeatureProvider.cs:32]: It was throwing before and actually all public types of a module must be assigned to at least one feature anyway. It would be an error if someone uses this with an internal type and a type that is not registered in DI. 
Therefore I think we should keep it throwing.

File: src/OrchardCore/OrchardCore/Extensions/Features/TypeFeatureProvider.cs (hunk 10 above)
Comment (gvkries) [reply src/OrchardCore/OrchardCore/Extensions/Features/TypeFeatureProvider.cs:47]: Callers are enumerating only once, one even with an `Any()` condition. `ToArray()` would be worse performance wise.

File: src/OrchardCore/OrchardCore.Abstractions/Extensions/Features/ITypeFeatureProvider.cs (hunk 2 above)
Comment (gvkries) [reply src/OrchardCore/OrchardCore.Abstractions/Extensions/Features/ITypeFeatureProvider.cs]: Done.

File: src/OrchardCore/OrchardCore.Abstractions/Extensions/Features/ITypeFeatureProvider.cs (hunk 3 above)
Comment (gvkries) [reply src/OrchardCore/OrchardCore.Abstractions/Extensions/Features/ITypeFeatureProvider.cs:14]: Done.

File: src/OrchardCore.Modules/OrchardCore.Roles/Controllers/AdminController.cs
[the remark this answers is not in the document]
Comment (gvkries) [reply src/OrchardCore.Modules/OrchardCore.Roles/Controllers/AdminController.cs:1]: Done.

**Comment by gvkries** [comment]
I have now removed all `FeatureAttributes` from several classes that should not require it anymore. Even if I couldn't find a problem, I'm a little bit nervous with about change and I hope this doesn't break anything. 

Especially, the removal of the attribute from some classes that were assigned to the application default feature, i.e. they had a `[Feature(Application.DefaultFeatureId)]` assigned, causes those to be logically moved to a different feature. This is true e.g. for some core provides (e.g. `CoreShapes` etc.).

I think this should not break anything, but maybe it is more safe to keep the attribute on those classes that have no specific feature assigned via a startup.

**Comment by Piedone** [comment]
Yeah, those that had a Feature attribute with a feature different than what they'll get now should rather remain. I have no idea what may depend on them but it's probably safe to assume that something can.

**Comment by gvkries** [comment]
You are right, I added the attribute back to the classes in question. Shouldn't have removed them in the first place. 😇

**Comment by Piedone** [comment]
Are you ready for another round of reviews?

**Comment by gvkries** [comment]
> Are you ready for another round of reviews?

Yes, I try to attend to the meeting tomorrow, if there is one.

File: src/OrchardCore/OrchardCore.Abstractions/Extensions/Features/ITypeFeatureProvider.cs [hunk 15]
```diff
@@ -13,36 +13,34 @@ public interface ITypeFeatureProvider
...
         /// </summary>
-        /// <param name="dependency"></param>
-        /// <returns></returns>
+        /// <remarks>
+        /// If a type has been registered for more than one feature,
+        /// <c>GetFeatureForDependency</c> returns the first feature that has
+        /// the same ID as the parent extension.
+        /// Use this method when you only need one feature of a module, such
+        /// as when applying migrations for the entire module as opposed to
+        /// functionality of individual features.
```
Comment (Piedone) [inline src/OrchardCore/OrchardCore.Abstractions/Extensions/Features/ITypeFeatureProvider.cs:27]:
```suggestion
        /// If a type has been registered for more than one feature,
        /// <see cref="GetFeatureForDependency"/> returns the first feature that has
        /// the same ID as the parent extension.
        /// Use this method when you only need one feature of a module, such
        /// as when applying migrations for the entire module as opposed to
        /// functionality of individual features.
```

File: src/OrchardCore.Modules/OrchardCore.Roles/Controllers/AdminController.cs [hunk 16]
```diff
@@ -241,9 +245,13 @@ private RoleEntry BuildRoleEntry(IRole role)
         private async Task<IDictionary<PermissionGroupKey, IEnumerable<Permission>>> GetInstalledPermissionsAsync()
         {
             var installedPermissions = new Dictionary<PermissionGroupKey, IEnumerable<Permission>>();
+            var enabledFeatures = await _shellFeaturesManager.GetEnabledFeaturesAsync();
             foreach (var permissionProvider in _permissionProviders)
             {
-                var feature = _typeFeatureProvider.GetFeatureForDependency(permissionProvider.GetType());
+                var feature = _typeFeatureProvider
+                    .GetFeaturesForDependency(permissionProvider.GetType())
+                    .LastOrDefault(feature => enabledFeatures.Any(enabledFeature => feature.Id == enabledFeature.Id));
```
Comment (Piedone) [inline src/OrchardCore.Modules/OrchardCore.Roles/Controllers/AdminController.cs:253]: Why is this necessary?

**Review by Piedone — CHANGES_REQUESTED** [review]

**Comment by MikeAlhayek** [comment]
@gvkries during the call, you mentioned there are minor things that you want to do to this PR. Once you are done, I'll do one last final review.

**Review by Piedone — APPROVED** [review]

**Comment by gvkries** [comment]
> @gvkries during the call, you mentioned there are minor things that you want to do to this PR. Once you are done, I'll do one last final review.

No sorry, I' done with this for now. I was just trying to explain that I'm not happy that some types are now assigned to two features by default. But I want to keep it this way, because I'm not sure if I break something if I change that. Basically all types are assigned to a feature that has the same ID as the parent extension. This will be the first feature entry and an additional more specific feature is added afterwards, if there is one. This way calling `GetFeatureForDependency` will return the same value as before.

So please go ahead and check it out 🙈

**Review by MikeAlhayek — APPROVED** [review]
I added a documentation in the release notes for awareness

**Comment by gvkries** [comment]
> I added a documentation in the release notes for awareness

@MikeAlhayek Many thanks!
